examples/plot_metric_examples.py

```diff
@@ -0,0 +1,410 @@
+# closer than two points in different labels.
+# 
+# Do keep in mind that we are doing this method because we know the labels
+# - we can actually create the constraints any way we want to depending on
+# the data!
+# 
+
+def create_constraints(labels):
+    import itertools
+    import random
+    
+    # aggregate indices of same class
+    zeros = np.where(Y==0)[0]
+    ones = np.where(Y==1)[0]
+    twos = np.where(Y==2)[0]
+    # make permutations of all those points in the same class
+    zeros_ = list(itertools.combinations(zeros, 2))
+    ones_ = list(itertools.combinations(ones, 2))
+    twos_ = list(itertools.combinations(twos, 2))
+    # put them together!
+    sim = np.array(zeros_ + ones_ + twos_)
+    
+    # similarily, put together indices in different classes
+    dis = []
+    for zero in zeros:
+        for one in ones:
+            dis.append((zero, one))
+        for two in twos:
+            dis.append((zero, two))
+    for one in ones:
+        for two in twos:
+            dis.append((one, two))
+            
+    # pick up just enough dissimilar examples as we have similar examples
+    dis = np.array(random.sample(dis, len(sim)))
+    
+    # return an array of pairs of indices of shape=(2*len(sim), 2), and the corresponding labels, array of shape=(2*len(sim))
+    # Each pair of similar points have a label of +1 and each pair of dissimilar points have a label of -1
+    return (np.vstack([np.column_stack([sim[:,0], sim[:,1]]), np.column_stack([dis[:,0], dis[:,1]])]),
+            np.concatenate([np.ones(len(sim)), -np.ones(len(sim))]))
+
+pairs, pairs_labels = create_constraints(Y)
+
+
+######################################################################
+# Now that we've created our constraints, let's see what it looks like!
+# 
+
+print(pairs)
+print(pairs_labels)
+
+
+######################################################################
+# Using our constraints, let's now train ITML again. We should keep in
+# mind that internally, ITML\_Supervised does pretty much the same thing
+# we are doing; I was just giving an example to better explain how the
+# constraints are structured.
+# 
+
+itml = metric_learn.ITML(preprocessor=X)
+itml.fit(pairs, pairs_labels)
+
+X_itml = itml.transform(X)
+
+plot(X_itml, Y)
+
+
+######################################################################
+# And that's the result of ITML after being trained on our manual
+# constraints! A bit different from our old result but not too different.
+# We can also notice that it might be better to rely on the randomised
+# algorithms under the hood to make our constraints if we are not very
+# sure how we want our transformed space to be.
+# 
+# RCA and SDML also have their own specific ways of taking in inputs -
+# it's worth one's while to poke around in the constraints.py file to see
+# how exactly this is going on.
+# 
+# This brings us to the end of this tutorial! Have fun Metric Learning :)
```

Comment:
maybe also add that metric-learn is compatible with sklearn and so we can easily do model selection, cross validation, scoring etc and refer to the doc for more details.

Reply:
I agree, done